corehq/apps/es/index/settings.py

```diff
@@ -0,0 +1,154 @@
+            if invalid_settings:
+                if collection is DEFAULT_TUNING_SETTINGS:
+                    where = f"{__name__}.DEFAULT_TUNING_SETTINGS"
+                else:
+                    where = "django.conf.settings.ES_SETTINGS"
+                msg = (f"Found invalid Elastic index tuning settings for "
+                       f"{where}[{settings_key!r}]: {invalid_settings}")
+                raise InvalidTuningSettingsError(msg)
+            for tunable in valid_tuning_keys:
+                try:
+                    value = settings_values[tunable]
+                except KeyError:
+                    continue
+                if value is REMOVE_SETTING:
+                    del tuning_settings[tunable]
+                else:
+                    tuning_settings[tunable] = value
+    return tuning_settings
+
+
+class InvalidTuningSettingsError(Exception):
+    pass
+
+
+class SettingsKeyCollection:
+    """An Enum-like collection of values used to reference Elastic index tuning
+    configurations in Django settings.
+
+    An ``enum.Enum`` subclass was not used because it is not well suited for
+    this task. That is to say: consumers of this type only wish to compare
+    *string values*, and converting values to/from ``str`` and "enum member"
+    would add unnecessary type cruft to the implementation without much (if any)
+    added value. While investigating this option, it also turns out that Python
+    does not have a ``StrEnum`` type until version 3.11.
+    """
+
+    @cached_property
+    def _values(self):
+        """Returns a set of class attribute values whose:
+        - attribute name does not start with an underscore (``_``)
+        - value is a string
+        """
+        values = set()
+        for key, value in type(self).__dict__.items():
+            if not key.startswith("_") and type(value) is str:
+                values.add(value)
+        return values
+
+    def __repr__(self):
+        return f"<{type(self).__name__} values={self._values}>"
+
+    def __contains__(self, value):
+        return value in self._values
+
+    def __iter__(self):
+        yield from self._values
+
+
+class _IndexSettingsKey(SettingsKeyCollection):
+    APPS = "hqapps"
+    CASES = "hqcases"
+    CASE_SEARCH = "case_search"
+    DOMAINS = "hqdomains"
+    FORMS = "xforms"
+    GROUPS = "hqgroups"
+    SMS = "smslogs"
+    USERS = "hqusers"
+
+
+class _IndexTuningKey(SettingsKeyCollection):
+    REPLICAS = "number_of_replicas"
+    SHARDS = "number_of_shards"
+
+
+IndexSettingsKey = _IndexSettingsKey()
+IndexTuningKey = _IndexTuningKey()
+
+
+# Allow removing settings from defaults by setting the value to this constant in
+# settings.ES_SETTINGS.
+# TODO: change this constant to an object that isn't actually a meaningful
+# (e.g. `__do_not_set__`) instead of `None` (because `None` is how you revert a
+# setting to the value the server is configured to use).
+REMOVE_SETTING = None
```

Comment:
Why is changing this left as a TODO?

Additionally, `None` is a common and easily referenced value. What do you think of using something a little harder to get by accident? (This doesn't need to be JSON-serializable, right? If it does, then `"__do_not_set__"` seems reasonable.)
```suggestion
REMOVE_SETTING = object()
```

Reply:
It _does_ need to be JSON-serializable (e.g. [elasticsearch.yml](https://github.com/dimagi/commcare-cloud/blob/master/environments/staging/elasticsearch.yml))